File: 09/claude-opus-4.5/go/main.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

type Point struct {
	x, y int64
}

type Edge struct {
	p1, p2 Point
}

func abs(n int64) int64 {
	if n < 0 {
		return -n
	}
	return n
}

func min64(a, b int64) int64 {
	if a < b {
		return a
	}
	return b
}

func max64(a, b int64) int64 {
	if a > b {
		return a
	}
	return b
}

func isOnSegment(px, py, x1, y1, x2, y2 int64) bool {
	if x1 == x2 {
		return px == x1 && py >= min64(y1, y2) && py <= max64(y1, y2)
	}
	if y1 == y2 {
		return py == y1 && px >= min64(x1, x2) && px <= max64(x1, x2)
	}
	return false
}

func isInsidePolygon(px, py int64, edges []Edge) bool {
	crossings := 0
	for _, e := range edges {
		y1, y2 := e.p1.y, e.p2.y
		x1, x2 := e.p1.x, e.p2.x
		if (y1 <= py && y2 > py) || (y2 <= py && y1 > py) {
			xIntersect := float64(x1) + float64(py-y1)/float64(y2-y1)*float64(x2-x1)
			if float64(px) < xIntersect {
				crossings++
			}
		}
	}
	return crossings%2 == 1
}

func isValidPoint(px, py int64, edges []Edge, tilesSet map[Point]bool) bool {
	if tilesSet[Point{px, py}] {
		return true
	}
	for _, e := range edges {
		if isOnSegment(px, py, e.p1.x, e.p1.y, e.p2.x, e.p2.y) {
			return true
		}
	}
	return isInsidePolygon(px, py, edges)
}

func direction(x1, y1, x2, y2, x3, y3 int64) float64 {
	return float64(x3-x1)*float64(y2-y1) - float64(x2-x1)*float64(y3-y1)
}

func segmentsIntersect(ax1, ay1, ax2, ay2, bx1, by1, bx2, by2 int64) bool {
	d1 := direction(bx1, by1, bx2, by2, ax1, ay1)
	d2 := direction(bx1, by1, bx2, by2, ax2, ay2)
	d3 := direction(ax1, ay1, ax2, ay2, bx1, by1)
	d4 := direction(ax1, ay1, ax2, ay2, bx2, by2)
	return ((d1 > 0 && d2 < 0) || (d1 < 0 && d2 > 0)) &&
		((d3 > 0 && d4 < 0) || (d3 < 0 && d4 > 0))
}

func getIntersection(ax1, ay1, ax2, ay2, bx1, by1, bx2, by2 int64) (float64, float64, bool) {
	d := float64(ax2-ax1)*float64(by2-by1) - float64(ay2-ay1)*float64(bx2-bx1)
	if math.Abs(d) < 1e-10 {
		return 0, 0, false
	}
	t := (float64(bx1-ax1)*float64(by2-by1) - float64(by1-ay1)*float64(bx2-bx1)) / d
	return float64(ax1) + t*float64(ax2-ax1), float64(ay1) + t*float64(ay2-ay1), true
}

func edgeCrossesRectInterior(x1, y1, x2, y2, minX, maxX, minY, maxY int64) bool {
	p1Inside := x1 >= minX && x1 <= maxX && y1 >= minY && y1 <= maxY
	p2Inside := x2 >= minX && x2 <= maxX && y2 >= minY && y2 <= maxY
	if p1Inside && p2Inside {
		return false
	}

	// Check each rect edge
	edges := [][4]int64{
		{minX, minY, maxX, minY}, // top
		{minX, maxY, maxX, maxY}, // bottom
		{minX, minY, minX, maxY}, // left
		{maxX, minY, maxX, maxY}, // right
	}

	for i, re := range edges {
		if segmentsIntersect(x1, y1, x2, y2, re[0], re[1], re[2], re[3]) {
			ix, iy, ok := getIntersection(x1, y1, x2, y2, re[0], re[1], re[2], re[3])
			if ok {
				if i < 2 { // horizontal edges
					if ix > float64(minX) && ix < float64(maxX) {
						return true
					}
				} else { // vertical edges
					if iy > float64(minY) && iy < float64(maxY) {
						return true
					}
				}
			}
		}
	}
	return false
}

func isRectangleValid(minX, maxX, minY, maxY int64, edges []Edge) bool {
	for _, e := range edges {
		if edgeCrossesRectInterior(e.p1.x, e.p1.y, e.p2.x, e.p2.y, minX, maxX, minY, maxY) {
			return false
		}
	}
	return true
}

func main() {
	file, _ := os.Open("input.txt")
	defer file.Close()

	var tiles []Point
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		parts := strings.Split(line, ",")
		x, _ := strconv.ParseInt(parts[0], 10, 64)
		y, _ := strconv.ParseInt(parts[1], 10, 64)
		tiles = append(tiles, Point{x, y})
	}

	// Part 1
	var maxArea int64 = 0
	for i := 0; i < len(tiles); i++ {
		for j := i + 1; j < len(tiles); j++ {
			width := abs(tiles[j].x-tiles[i].x) + 1
			height := abs(tiles[j].y-tiles[i].y) + 1
			area := width * height
			if area > maxArea {
				maxArea = area
			}
		}
	}
	fmt.Printf("Day 09 Part 1: %d\n", maxArea)

	// Part 2
	edges := make([]Edge, len(tiles))
	for i := 0; i < len(tiles); i++ {
		edges[i] = Edge{tiles[i], tiles[(i+1)%len(tiles)]}
	}

	tilesSet := make(map[Point]bool)
	for _, t := range tiles {
		tilesSet[t] = true
	}

	var maxArea2 int64 = 0
	for i := 0; i < len(tiles); i++ {
		for j := i + 1; j < len(tiles); j++ {
			t1, t2 := tiles[i], tiles[j]
			minX, maxX := min64(t1.x, t2.x), max64(t1.x, t2.x)
			minY, maxY := min64(t1.y, t2.y), max64(t1.y, t2.y)

			corners := []Point{{minX, minY}, {minX, maxY}, {maxX, minY}, {maxX, maxY}}
			allValid := true
			for _, c := range corners {
				if !isValidPoint(c.x, c.y, edges, tilesSet) {
					allValid = false
					break
				}
			}
			if allValid && isRectangleValid(minX, maxX, minY, maxY, edges) {
				area := (maxX - minX + 1) * (maxY - minY + 1)
				if area > maxArea2 {
					maxArea2 = area
				}
			}
		}
	}
	fmt.Printf("Day 09 Part 2: %d\n", maxArea2)
}

File: 09/claude-opus-4.5/rust/src/main.rs
use std::fs;
use std::collections::HashSet;

fn main() {
    let input = fs::read_to_string("input.txt").expect("Failed to read input");

    let tiles: Vec<(i64, i64)> = input
        .lines()
        .filter(|line| !line.trim().is_empty())
        .map(|line| {
            let parts: Vec<&str> = line.split(',').collect();
            (
                parts[0].trim().parse().unwrap(),
                parts[1].trim().parse().unwrap(),
            )
        })
        .collect();

    // Part 1
    let mut max_area: i64 = 0;
    for i in 0..tiles.len() {
        for j in (i + 1)..tiles.len() {
            let width = (tiles[j].0 - tiles[i].0).abs() + 1;
            let height = (tiles[j].1 - tiles[i].1).abs() + 1;
            let area = width * height;
            if area > max_area {
                max_area = area;
            }
        }
    }
    println!("Day 09 Part 1: {}", max_area);

    // Part 2
    let edges: Vec<((i64, i64), (i64, i64))> = (0..tiles.len())
        .map(|i| (tiles[i], tiles[(i + 1) % tiles.len()]))
        .collect();

    let tiles_set: HashSet<(i64, i64)> = tiles.iter().cloned().collect();

    let mut max_area2: i64 = 0;
    for i in 0..tiles.len() {
        for j in (i + 1)..tiles.len() {
            let (t1, t2) = (tiles[i], tiles[j]);
            let min_x = t1.0.min(t2.0);
            let max_x = t1.0.max(t2.0);
            let min_y = t1.1.min(t2.1);
            let max_y = t1.1.max(t2.1);

            let corners = [(min_x, min_y), (min_x, max_y), (max_x, min_y), (max_x, max_y)];
            let all_valid = corners.iter().all(|&(cx, cy)| is_valid_point(cx, cy, &edges, &tiles_set));

            if all_valid && is_rectangle_valid(min_x, max_x, min_y, max_y, &edges) {
                let area = (max_x - min_x + 1) * (max_y - min_y + 1);
                if area > max_area2 {
                    max_area2 = area;
                }
            }
        }
    }
    println!("Day 09 Part 2: {}", max_area2);
}

fn is_on_segment(px: i64, py: i64, x1: i64, y1: i64, x2: i64, y2: i64) -> bool {
    if x1 == x2 {
        px == x1 && py >= y1.min(y2) && py <= y1.max(y2)
    } else if y1 == y2 {
        py == y1 && px >= x1.min(x2) && px <= x1.max(x2)
    } else {
        false
    }
}

fn is_inside_polygon(px: i64, py: i64, edges: &[((i64, i64), (i64, i64))]) -> bool {
    let mut crossings = 0;
    for &((x1, y1), (x2, y2)) in edges {
        if (y1 <= py && y2 > py) || (y2 <= py && y1 > py) {
            let x_intersect = x1 as f64 + (py - y1) as f64 / (y2 - y1) as f64 * (x2 - x1) as f64;
            if (px as f64) < x_intersect {
                crossings += 1;
            }
        }
    }
    crossings % 2 == 1
}

fn is_valid_point(px: i64, py: i64, edges: &[((i64, i64), (i64, i64))], tiles_set: &HashSet<(i64, i64)>) -> bool {
    if tiles_set.contains(&(px, py)) {
        return true;
    }
    for &((x1, y1), (x2, y2)) in edges {
        if is_on_segment(px, py, x1, y1, x2, y2) {
            return true;
        }
    }
    is_inside_polygon(px, py, edges)
}

fn direction(x1: i64, y1: i64, x2: i64, y2: i64, x3: i64, y3: i64) -> f64 {
    (x3 - x1) as f64 * (y2 - y1) as f64 - (x2 - x1) as f64 * (y3 - y1) as f64
}

fn segments_intersect(ax1: i64, ay1: i64, ax2: i64, ay2: i64, bx1: i64, by1: i64, bx2: i64, by2: i64) -> bool {
    let d1 = direction(bx1, by1, bx2, by2, ax1, ay1);
    let d2 = direction(bx1, by1, bx2, by2, ax2, ay2);
    let d3 = direction(ax1, ay1, ax2, ay2, bx1, by1);
    let d4 = direction(ax1, ay1, ax2, ay2, bx2, by2);
    ((d1 > 0.0 && d2 < 0.0) || (d1 < 0.0 && d2 > 0.0)) &&
    ((d3 > 0.0 && d4 < 0.0) || (d3 < 0.0 && d4 > 0.0))
}

fn get_intersection(ax1: i64, ay1: i64, ax2: i64, ay2: i64, bx1: i64, by1: i64, bx2: i64, by2: i64) -> Option<(f64, f64)> {
    let d = (ax2 - ax1) as f64 * (by2 - by1) as f64 - (ay2 - ay1) as f64 * (bx2 - bx1) as f64;
    if d.abs() < 1e-10 {
        return None;
    }
    let t = ((bx1 - ax1) as f64 * (by2 - by1) as f64 - (by1 - ay1) as f64 * (bx2 - bx1) as f64) / d;
    Some((ax1 as f64 + t * (ax2 - ax1) as f64, ay1 as f64 + t * (ay2 - ay1) as f64))
}

fn edge_crosses_rect_interior(x1: i64, y1: i64, x2: i64, y2: i64, min_x: i64, max_x: i64, min_y: i64, max_y: i64) -> bool {
    let p1_inside = x1 >= min_x && x1 <= max_x && y1 >= min_y && y1 <= max_y;
    let p2_inside = x2 >= min_x && x2 <= max_x && y2 >= min_y && y2 <= max_y;
    if p1_inside && p2_inside {
        return false;
    }

    let rect_edges = [
        (min_x, min_y, max_x, min_y),
        (min_x, max_y, max_x, max_y),
        (min_x, min_y, min_x, max_y),
        (max_x, min_y, max_x, max_y),
    ];

    for (i, &(rx1, ry1, rx2, ry2)) in rect_edges.iter().enumerate() {
        if segments_intersect(x1, y1, x2, y2, rx1, ry1, rx2, ry2) {
            if let Some((ix, iy)) = get_intersection(x1, y1, x2, y2, rx1, ry1, rx2, ry2) {
                if i < 2 {
                    if ix > min_x as f64 && ix < max_x as f64 {
                        return true;
                    }
                } else if iy > min_y as f64 && iy < max_y as f64 {
                    return true;
                }
            }
        }
    }
    false
}

fn is_rectangle_valid(min_x: i64, max_x: i64, min_y: i64, max_y: i64, edges: &[((i64, i64), (i64, i64))]) -> bool {
    for &((x1, y1), (x2, y2)) in edges {
        if edge_crosses_rect_interior(x1, y1, x2, y2, min_x, max_x, min_y, max_y) {
            return false;
        }
    }
    true
}